File: DeepSpace-Ruby/lib/Weapon.rb
# encoding:utf-8

require_relative "WeaponToUI"
require_relative "WeaponType"

module Deepspace
class Weapon
#
# ──────────────────────────────────────────────────────────── CONSTRUCTORES ─────
#

    def initialize (nombre, tipo, usos)
        @name = nombre
        @type = tipo
        @uses = usos
    end

    def self.newCopy(s)
        new(s.name, s.type, s.uses)
    end

#
# ─────────────────────────────────────────────────────────────── INTERFACES ─────
#

    attr_reader :name, :type, :uses

    def power
        @type.power
    end

    def useIt
        if @uses > 0
            @uses -= 1
            power
        else
            1.0
        end
    end

    def getUIversion
        WeaponToUI.new(self)
    end

    def to_s
        getUIversion.to_s
    end
end
end

File: DeepSpace-Java/src/deepspace/Weapon.java
package deepspace;

//
// ──────────────────────────────────────────────────────────────── I ──────────
//   :::::: C L A S E   W E A P O N : :  :   :    :     :        :          :
// ──────────────────────────────────────────────────────────────────────────
//

class Weapon implements Copyable <Weapon>, CombatElement {
    private String name;
    private WeaponType type;
    private int uses;

    Weapon (String nombre, WeaponType tipo, int usos) {
        name = nombre;
        type = tipo;
        uses = usos;
    }

    Weapon (Weapon otro) {
        name = otro.name;
        type = otro.type;
        uses = otro.uses;
    }

    @Override
    public Weapon copy() {
        return new Weapon(this);
    }

    public WeaponType getType () {
        return type;
    }

    public int getUses () {
        return uses;
    }

    public float power() {
        return type.getPower();
    }

    public float useIt() {
        if (uses > 0) {
            uses--;
            return this.power();
        }
        else {
            return 1.0f;
        }
    }

    WeaponToUI getUIversion () {
        return new WeaponToUI(this);
    }

    @Override
    public String toString() {
        return getUIversion.toString;
    }
}
